### PR TITLE
Space | Nora & Sara

### PR DESCRIPTION
**Pull request by thenora**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? |  It was really hard. We didn't understand what large sections of the code meant. But we took notes, and followed up to ask questions about things like "must_respond_to" in tests. We continued to learn more about it as we progressed in the program. |
What _inheritance_ relations exist between classes? |  Both driver, passenger, and trips inherit from CsvRecord |
What _composition_ relations exist between classes? | *********** N|
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |  What do we do when the cost of the ride is less than the fee of $1.65? We thought about the driver's needs, and the company's needs, and decided to not charge the fee at all, and give the driver 80% rather than give the driver nothing. We didn't want drivers to not get paid at all, but also didn't want to not pay the company. Plus we thought about the difficulty of different methods for writing the code. |
Give an example of a _template method_ that you implemented for this assignment |  from_csv method was a template method we used and changed for the related classes. |
Give an example of a _nominal test_ that you wrote for this assignment |  For the total_revenue method on Driver, we wrote a test to see if it correctly handled the trip cost of a trip less than $1.65. |
Give an example of an _edge case test_ that you wrote for this assignment | for the trip_test we wrote "Raises an error if the end time is before the start time" |
What is a concept that you gained more clarity on as you worked on this assignment | We learned a lot more about classes, and inheritance, and how nil is handled in equations. |

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -1,66 +1,86 @@
...
       @cost = cost
       @rating = rating
 
-      if @rating > 5 || @rating < 1
-        raise ArgumentError.new("Invalid rating #{@rating}")
+      if @rating != nil
+        if @rating > 5 || @rating < 1
+          raise ArgumentError.new("Invalid rating #{@rating}")
+        end
+      end
```
Comment (kaidamasaki) [inline lib/trip.rb]: Generally we leave off the `!= nil` in Ruby.  Also, you can use a compound conditional here instead of nested `if`s:

```suggestion
      if @rating && (@rating > 5 || @rating < 1)
        raise ArgumentError.new("Invalid rating #{@rating}")
      end
```

File: lib/trip.rb [hunk 2]
```diff
@@ -1,66 +1,86 @@
...
+
+      if @end_time != nil
+        raise ArgumentError.new("The start time should be before the end time") if @end_time < @start_time
       end
```
Comment (kaidamasaki) [inline lib/trip.rb]: This can be simplified like above:

```suggestion
      raise ArgumentError.new("The start time should be before the end time") if @end_time && @end_time < @start_time
```

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -1,26 +1,60 @@
...
       @trips = Trip.load_all(directory: directory)
+      @drivers = Driver.load_all(directory: directory)
       connect_trips
     end
 
     def find_passenger(id)
       Passenger.validate_id(id)
-      return @passengers.find { |passenger| passenger.id == id }
+      passenger = @passengers.find { |passenger| passenger.id == id }
+      return passenger
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb]: I'm unclear why you decided to assign this to a variable.  The previous line worked.

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -1,26 +1,60 @@
...
+
+    def find_driver(id)
+      Driver.validate_id(id)
+      return @drivers.find { |driver| driver.id == id }
+    end
+
+    def request_trip(passenger_id)
+      available_driver = @drivers.find { |driver| driver.status == :AVAILABLE }
+
+      raise ArgumentError.new("There are no drivers available") if available_driver == nil
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb]: This should be a [custom error](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/custom-exceptions.md), not an argument error. 

There's nothing wrong with the argument itself (changing the `passenger_id` won't help).

For example you could create an error class named `NoDriverAvailableError`:
```suggestion
      raise NoDriverAvailableError.new("There are no drivers available") if available_driver == nil
```

**Comment by kaidamasaki** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | ✔️
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | No edge case test.
Practices inheritence. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | ✔️
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 6+ in Code Review && 2+ in Functional Requirements | ✔️
| Red (Not at Standard) | 0-5 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | ✔️

### Additional Feedback



I'm not sure what "*********** N" means in your comprehension questions.


## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Well done!  Here are a few small ways you can clean up your code even more.  😄
